File: csharp-dotnet/src/Architecture.Core/Domain/Aggregates/IAggregateRoot.cs
using System.Collections.Generic;
using Architecture.Core.Domain.Entities;
using Architecture.Core.Domain.Events;

namespace Architecture.Core.Domain.Aggregates;

/// <summary>
/// Defines the contract for aggregate roots that manage consistency boundaries and domain events.
/// Aggregate roots are the only entities that can be directly accessed from outside the aggregate boundary.
/// </summary>
/// <typeparam name="TId">The type of the aggregate root's identifier, constrained to reference types.</typeparam>
public interface IAggregateRoot<TId> : IEntity<TId>
    where TId : class
{
    /// <summary>
    /// Gets the version number used for optimistic concurrency control.
    /// The version is incremented each time the aggregate is persisted.
    /// </summary>
    long Version { get; }

    /// <summary>
    /// Gets the collection of domain events that have been raised by this aggregate.
    /// Events represent important business occurrences that other parts of the system may need to react to.
    /// </summary>
    IReadOnlyCollection<IDomainEvent> Events { get; }

    /// <summary>
    /// Clears all uncommitted domain events from this aggregate.
    /// Typically called by the infrastructure after events have been published or persisted.
    /// </summary>
    void ClearEvents();
}

File: typescript-nodejs/architecture-core/src/domain/interfaces/i-aggregate-root.ts
import { IEntity } from './i-entity';
import { IDomainEvent } from './i-domain-event';

/**
 * Interface for aggregate roots.
 * An aggregate root is the entry point to an aggregate and the only object that external objects
 * should hold references to. It maintains consistency boundaries and collects domain events.
 *
 * @template TId The type of the aggregate's identifier
 */
export interface IAggregateRoot<TId extends object> extends IEntity<TId> {
  /** Version number for optimistic concurrency control */
  readonly version: number;

  /** Collection of domain events that occurred within this aggregate */
  readonly events: readonly IDomainEvent[];

  /**
   * Clears all collected domain events.
   * This is typically called after events have been published by the infrastructure layer.
   */
  clearEvents(): void;
}